Replace remaining sprintf with snprintf

laanwj opened the pull request:
Use of `sprintf` is seen as a red flag as many of its uses are insecure. OpenBSD warns about it while compiling, and some modern platforms, e.g. [cloudlibc from cloudabi](https://github.com/NuxiNL/cloudlibc) don't even provide it anymore.

Although our uses of these functions are secure, it can't hurt to replace them anyway. There are only 4 occurences of which 3 in the tests.

laanwj committed:
test: Replace remaining sprintf with snprintf

Use of `sprintf` is seen as a red flag as many of its uses are insecure.
OpenBSD warns about it while compiling, and some modern platforms, e.g.
[cloudlibc from cloudabi](https://github.com/NuxiNL/cloudlibc) don't
even provide it anymore.

Although our uses of these functions are secure, it can't hurt to
replace them anyway. There are only 3 occurences left, all in the
tests.

gmaxwell reviewed (approved):
utACK.

paveljanik commented:
ACK https://github.com/bitcoin/bitcoin/pull/9867/commits/d9b6b874c955cdd0e174b8f1e9582773e15c7cd9

JeremyRubin commented on src/uint256.cpp:
```diff
@@ -22,7 +22,7 @@ std::string base_blob<BITS>::GetHex() const
     char psz[sizeof(data) * 2 + 1];
     for (unsigned int i = 0; i < sizeof(data); i++)
-        sprintf(psz + i * 2, "%02x", data[sizeof(data) - i - 1]);
+        snprintf(psz + i * 2, 3, "%02x", data[sizeof(data) - i - 1]);
```
This is a little bit sketchy. Could it not safely be 2 here, not 3?

kallewoof replied:
NUL term needs a 3rd byte.

dcousens replied:
For others http://www.cplusplus.com/reference/cstdio/snprintf/

> A terminating null character is automatically appended after the content written.

Hence,  2 characters + 1 null character.

gmaxwell replied:
It's correct, and it's also why snprintf isn't an especially safe function call.

laanwj replied:
> It's correct, and it's also why snprintf isn't an especially safe function call. (replaces one class of issue with another more subtle one)
Indeed, that's why we never use it. The go-to function we have for this is "strprintf". However, it's kind of high overhead for formatting two hex characters. Especially as this function is called a lot.

I'll just replace this with use of `HexDigit` instead.

TheBlueMatt commented:
Concept ACK for general sanity, though I'm pretty sure there is no reason we should go out of our way to support libcs that are deliberately broken.

laanwj commented:
> Concept ACK for general sanity, though I'm pretty sure there is no reason we should go out of our way to support libcs that are deliberately broken.

Somewhat tend to agree, though there is something to be said for deprecating unsafe functions that have caused so much pain since the 80's. They do the same with system calls, trying to keep the API as minimalist as possible.

laanwj committed:
uint256: replace sprintf with HexStr and reverse-iterator

Instead of calling sprintf for every byte, format the hex bytes
ourselves by help of HexStr and a reverse_iterator.

laanwj commented:
Force-pushed into two commits:
- uint256: replace use of sprintf with per-digit hex conversion
- test: Replace remaining sprintf with snprintf

dcousens commented on src/uint256.cpp:
```diff
@@ -20,10 +20,15 @@ base_blob<BITS>::base_blob(const std::vector<unsigned char>& vch)
+    static const char *hex_chars = "0123456789abcdef";
+    std::string retval;
+    retval.reserve(sizeof(data) * 2);
+    for (unsigned int i = 0; i < sizeof(data); i++) {
```
why not `size_t`?

laanwj replied:
Meh. For loop counters it's generally enough to use a 32-bit type. Especially if you know you're only going to count to 64, ever.

dcousens commented on src/uint256.cpp:
```diff
@@ -20,10 +20,15 @@ base_blob<BITS>::base_blob(const std::vector<unsigned char>& vch)
+    for (unsigned int i = 0; i < sizeof(data); i++) {
+        uint8_t value = data[sizeof(data) - i - 1];
+        retval.push_back(hex_chars[value >> 4]);
+        retval.push_back(hex_chars[value & 0xf]);
```
couldn't we just do:

``` cpp
retval.push_back(hex_chars[value & 0xf0]);
retval.push_back(hex_chars[value & 0x0f]);
```

It would be way clearer...

laanwj replied:
But incorrect.

JeremyRubin replied:
nit: maybe emplace rather than push (potentially saves an initialization)
@dcousens can you not delete your feedback? I had to dig in my emails to see that I wasn't repeating your feedback.

laanwj replied:
Do you really think there's anything to be gained by emplacing a char? I'd say something has to be really wrong for that to make a difference.

laanwj replied:
Anyhow I have a better solution based on reverse_iterator coming, so no need to discuss this further.

sipa replied:
I'm pretty sure that the copy constructor and move constructor for a char are the same :D, so emplace wouldn't make a difference.

dcousens replied:
@JeremyRubin I made a woefully incorrect suggestion in regards to the bitmasking.

``` cpp
retval.push_back(hex_chars[value & 0xf0]);
retval.push_back(hex_chars[value & 0x0f]);
```

laanwj replied:
@dcousens no worries we all have our [dumb moments](https://github.com/bitcoin/bitcoin/pull/9871#issuecomment-282645504).

laanwj commented on src/uint256.cpp:
```diff
@@ -20,10 +20,15 @@ base_blob<BITS>::base_blob(const std::vector<unsigned char>& vch)
 template <unsigned int BITS>
 std::string base_blob<BITS>::GetHex() const
```
Possibly we could even use HexStr here with a reverse_iterator, though I don't know how to do that for a C-style array. All the variations I can think of are much worse than just replicating this functionality here.

JeremyRubin replied:
```c++
#include "stdio.h"
#include <iterator>
int main() {
    char a[2] = {'a', 'b'};
    auto start = std::reverse_iterator<char *>(&a[2]);
    auto end = std::reverse_iterator<char *>(&a[0]);
    while (start != end)
        printf("%c", *(start++));
}
```
Outputs:
> ba

laanwj replied:
Yes, thanks, I'm aware, see the new version of this patch.

JeremyRubin replied:
ha -- and here I was thinking you coded it after I commented and I was like "wow, @laanwj is a god amongst men in code response times" ;)

laanwj replied:
Yes, I implemented it in minus 3 minutes! Truly godlike :-)

laanwj commented:
There you go, a one-liner now.

JeremyRubin commented:
utACK 19cafc6

practicalswift commented:
utACK 19cafc6

sipa commented:
utACK

gmaxwell reviewed (approved):
re-utACK.

kallewoof commented:
utACK 19cafc6

jonasschnelli reviewed (approved):
utACK 19cafc6239abd14f2b9c3d883dc7df0472cac52b
